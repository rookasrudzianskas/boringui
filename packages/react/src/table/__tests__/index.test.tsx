import React from "react";
import {mount} from "enzyme";

import Table from "../index";

const STITCHES_FACTOR = 2;

const columns = [
  {name: "Foo", key: "foo"},
  {name: "Bar", key: "bar"},
  {name: "Baz", key: "baz"},
];

let items = [
  {test: "Test 1", foo: "Foo 1", bar: "Bar 1", yay: "Yay 1", baz: "Baz 1"},
  {test: "Test 2", foo: "Foo 2", bar: "Bar 2", yay: "Yay 2", baz: "Baz 2"},
];

describe("Table", () => {
  it("should render correctly", () => {
    const wrapper = mount(
      <Table aria-label="Test example table">
        <Table.Header>
          <Table.Column>NAME</Table.Column>
          <Table.Column>TYPE</Table.Column>
          <Table.Column>DATE MODIFIED</Table.Column>
        </Table.Header>
        <Table.Body>
          <Table.Row key="1">
            <Table.Cell>Games</Table.Cell>
            <Table.Cell>File folder</Table.Cell>
            <Table.Cell>6/7/2020</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should render a static table", () => {
    const wrapper = mount(
      <Table aria-label="Static Table">
        <Table.Header>
          <Table.Column>Foo</Table.Column>
          <Table.Column>Bar</Table.Column>
          <Table.Column>Baz</Table.Column>
        </Table.Header>
        <Table.Body>
          <Table.Row>
            <Table.Cell>Foo 1</Table.Cell>
            <Table.Cell>Bar 1</Table.Cell>
            <Table.Cell>Baz 1</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );

    const table = wrapper.find('[role="grid"]').at(0);

    expect(table.props()["aria-label"]).toBe("Static Table");
    expect(table.find('[role="rowgroup"]').at(0).length).not.toBeNull();
    expect(table.find('[role="rowgroup"]').at(1).length).not.toBeNull();

    // get all headers
    const headers = table.find('[role="columnheader"]');

    expect(headers.length).toBe(3 * STITCHES_FACTOR); // is multiplied by 2 because of the stitches components

    headers.forEach((header) => {
      expect(header.props()["aria-sort"]).toBeFalsy();
      expect(header.props()["aria-describedby"]).toBeFalsy();
    });

    expect(headers.at(0 * STITCHES_FACTOR).text()).toBe("Foo");
    expect(headers.at(1 * STITCHES_FACTOR).text()).toBe("Bar");
    expect(headers.at(2 * STITCHES_FACTOR).text()).toBe("Baz");

    const bodyRowGroup = table.find('[role="rowgroup"]').at(1 * STITCHES_FACTOR);

    // get all rows
    const rows = bodyRowGroup.find('[role="row"]');

    expect(rows.length).toBe(2);

    // get body first cell (rowheader)
    const bodyFirstCell = rows.at(0).find('[role="rowheader"]').at(0);

    expect(bodyFirstCell.text()).toBe("Foo 1");

    // get all cells
    const cells = rows.at(0).find('[role="gridcell"]');

    expect(cells.length).toBe(2 * STITCHES_FACTOR);
  });

  it("should render a table with selection", () => {
    const wrapper = mount(
      <Table aria-label="Table with selection" selectionMode="multiple">
        <Table.Header>
          <Table.Column>Foo</Table.Column>
          <Table.Column>Bar</Table.Column>
          <Table.Column>Baz</Table.Column>
        </Table.Header>
        <Table.Body>
          <Table.Row>
            <Table.Cell>Foo 1</Table.Cell>
            <Table.Cell>Bar 1</Table.Cell>
            <Table.Cell>Baz 1</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );

    const table = wrapper.find('[role="grid"]').at(0);

    expect(table.props()["aria-label"]).toBe("Table with selection");
    expect(table.props()["aria-multiselectable"]).toBe("true");

    const bodyRowGroup = table.find('[role="rowgroup"]').at(1 * STITCHES_FACTOR);

    // select all checkbox
    let checkbox = table.find('[type="checkbox"]').at(0);

    expect(checkbox.props()["aria-label"]).toBe("Select All");

    // get all rows
    const rows = bodyRowGroup.find('[role="row"]');

    expect(rows.length).toBe(1 * STITCHES_FACTOR);

    // get body first cell (rowheader)
    const bodyFirstCell = rows.at(0).find('[role="rowheader"]').at(0);

    expect(bodyFirstCell.text()).toBe("Foo 1");

    const firstRowCheckbox = rows.at(0).find('[type="checkbox"]').at(0);

    expect(firstRowCheckbox.props()["aria-label"]).toBe("Select");
  });

  it("should render dynamic table", () => {
    const wrapper = mount(
      <Table aria-label="Dynamic Table">
        <Table.Header columns={columns}>
          {(column) => <Table.Column>{column.name}</Table.Column>}
        </Table.Header>
        <Table.Body items={items}>
          {(item: any) => (
            <Table.Row key={item.test}>
              <Table.Cell>{item.test}</Table.Cell>
              <Table.Cell>{item.foo}</Table.Cell>
              <Table.Cell>{item.bar}</Table.Cell>
            </Table.Row>
          )}
        </Table.Body>
      </Table>,
    );

    const table = wrapper.find('[role="grid"]').at(0);

    expect(table.props()["aria-label"]).toBe("Dynamic Table");

    const bodyRowGroup = table.find('[role="rowgroup"]').at(1 * STITCHES_FACTOR);

    // get all rows
    const rows = bodyRowGroup.find('[role="row"]');

    expect(rows.length).toBe(2 * STITCHES_FACTOR);

    // get body first cell (rowheader)
    const bodyFirstCell = rows.at(0).find('[role="rowheader"]').at(0);

    expect(bodyFirstCell.text()).toBe("Test 1");

    // get all cells
    const cells = rows.at(0).find('[role="gridcell"]');

    expect(cells.length).toBe(2 * STITCHES_FACTOR);
  });

  it("should render a dynamic table with selection", () => {
    const wrapper = mount(
      <Table aria-label="Dynamic Table with selection" selectionMode="multiple">
        <Table.Header columns={columns}>
          {(column) => <Table.Column>{column.name}</Table.Column>}
        </Table.Header>
        <Table.Body items={items}>
          {(item: any) => (
            <Table.Row key={item.test}>
              <Table.Cell>{item.test}</Table.Cell>
              <Table.Cell>{item.foo}</Table.Cell>
              <Table.Cell>{item.bar}</Table.Cell>
            </Table.Row>
          )}
        </Table.Body>
      </Table>,
    );

    const table = wrapper.find('[role="grid"]').at(0);

    expect(table.props()["aria-label"]).toBe("Dynamic Table with selection");
    expect(table.props()["aria-multiselectable"]).toBe("true");

    const bodyRowGroup = table.find('[role="rowgroup"]').at(1 * STITCHES_FACTOR);

    // select all checkbox
    let checkbox = table.find('[type="checkbox"]').at(0);

    expect(checkbox.props()["aria-label"]).toBe("Select All");

    // get all rows
    const rows = bodyRowGroup.find('[role="row"]');

    expect(rows.length).toBe(2 * STITCHES_FACTOR);

    // get body first cell (rowheader)
    const bodyFirstCell = rows.at(0).find('[role="rowheader"]').at(0);

    expect(bodyFirstCell.text()).toBe("Test 1");

    const firstRowCheckbox = rows.at(0).find('[type="checkbox"]').at(0);

    expect(firstRowCheckbox.props()["aria-label"]).toBe("Select");
  });

  it("should render a static table correctly with sorting", () => {
    const wrapper = mount(
      <Table aria-label="Test sorting table">
        <Table.Header>
          <Table.Column>NAME</Table.Column>
          <Table.Column allowsSorting>TYPE</Table.Column>
          <Table.Column allowsSorting>DATE MODIFIED</Table.Column>
        </Table.Header>
        <Table.Body>
          <Table.Row key="1">
            <Table.Cell>Games</Table.Cell>
            <Table.Cell>File folder</Table.Cell>
            <Table.Cell>6/7/2020</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );

    const table = wrapper.find('[role="grid"]').at(0);

    expect(table.props()["aria-label"]).toBe("Test sorting table");

    // get column headers
    const columnHeaders = table.find('[role="columnheader"]');

    expect(columnHeaders.length).toBe(3 * STITCHES_FACTOR);

    // check first column header
    const firstColumnHeader = columnHeaders.at(0 * STITCHES_FACTOR);

    expect(firstColumnHeader.text()).toBe("NAME");
    expect(firstColumnHeader.props()["aria-sort"]).toBeFalsy();

    // check second column header
    const secondColumnHeader = columnHeaders.at(1 * STITCHES_FACTOR);

    expect(secondColumnHeader.text()).toBe("TYPE");
    expect(secondColumnHeader.props()["aria-sort"]).toBe("none");

    // check third column header
    const thirdColumnHeader = columnHeaders.at(2 * STITCHES_FACTOR);

    expect(thirdColumnHeader.text()).toBe("DATE MODIFIED");
    expect(thirdColumnHeader.props()["aria-sort"]).toBe("none");

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should set the proper aria-sort on an ascending sorted column header", () => {
    const wrapper = mount(
      <Table
        aria-label="Test sorting table"
        sortDescriptor={{column: "type", direction: "ascending"}}
      >
        <Table.Header>
          <Table.Column key="name">NAME</Table.Column>
          <Table.Column key="type" allowsSorting>
            TYPE
          </Table.Column>
          <Table.Column key="date_modified" allowsSorting>
            DATE MODIFIED
          </Table.Column>
        </Table.Header>
        <Table.Body>
          <Table.Row key="1">
            <Table.Cell>Games</Table.Cell>
            <Table.Cell>File folder</Table.Cell>
            <Table.Cell>6/7/2020</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );

    const table = wrapper.find('[role="grid"]').at(0);

    expect(table.props()["aria-label"]).toBe("Test sorting table");

    // get column headers
    const columnHeaders = table.find('[role="columnheader"]');

    expect(columnHeaders.length).toBe(3 * STITCHES_FACTOR);

    // check first column header
    const firstColumnHeader = columnHeaders.at(0 * STITCHES_FACTOR);

    expect(firstColumnHeader.text()).toBe("NAME");
    expect(firstColumnHeader.props()["aria-sort"]).toBeFalsy();

    // check second column header
    const secondColumnHeader = columnHeaders.at(1 * STITCHES_FACTOR);

    expect(secondColumnHeader.text()).toBe("TYPE");
    expect(secondColumnHeader.props()["aria-sort"]).toBe("ascending");

    // check third column header
    const thirdColumnHeader = columnHeaders.at(2 * STITCHES_FACTOR);

    expect(thirdColumnHeader.text()).toBe("DATE MODIFIED");
    expect(thirdColumnHeader.props()["aria-sort"]).toBe("none");
  });

  it("should set the proper aria-sort on an descending sorted column header", () => {
    const wrapper = mount(
      <Table
        aria-label="Test sorting table"
        sortDescriptor={{column: "type", direction: "descending"}}
      >
        <Table.Header>
          <Table.Column key="name">NAME</Table.Column>
          <Table.Column key="type" allowsSorting>
            TYPE
          </Table.Column>
          <Table.Column key="date_modified" allowsSorting>
            DATE MODIFIED
          </Table.Column>
        </Table.Header>
        <Table.Body>
          <Table.Row key="1">
            <Table.Cell>Games</Table.Cell>
            <Table.Cell>File folder</Table.Cell>
            <Table.Cell>6/7/2020</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );

    const table = wrapper.find('[role="grid"]').at(0);

    expect(table.props()["aria-label"]).toBe("Test sorting table");

    // get column headers
    const columnHeaders = table.find('[role="columnheader"]');

    expect(columnHeaders.length).toBe(3 * STITCHES_FACTOR);

    // check first column header
    const firstColumnHeader = columnHeaders.at(0 * STITCHES_FACTOR);

    expect(firstColumnHeader.text()).toBe("NAME");
    expect(firstColumnHeader.props()["aria-sort"]).toBeFalsy();

    // check second column header
    const secondColumnHeader = columnHeaders.at(1 * STITCHES_FACTOR);

    expect(secondColumnHeader.text()).toBe("TYPE");
    expect(secondColumnHeader.props()["aria-sort"]).toBe("descending");

    // check third column header
    const thirdColumnHeader = columnHeaders.at(2 * STITCHES_FACTOR);

    expect(thirdColumnHeader.text()).toBe("DATE MODIFIED");
    expect(thirdColumnHeader.props()["aria-sort"]).toBe("none");
  });
});
