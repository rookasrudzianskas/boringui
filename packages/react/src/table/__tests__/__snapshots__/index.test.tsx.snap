// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render a static table correctly with sorting 1`] = `"<div class=\\"nextui-c-hWZRae nextui-c-hWZRae-gikTHb-shadow-true nextui-c-hWZRae-bzEnZV-borderWeight-normal nextui-c-hWZRae-iPJLV-css nextui-table-container\\"><table class=\\"nextui-c-fcXuxO nextui-c-fcXuxO-fyxmJu-color-primary nextui-c-fcXuxO-fPgLZC-lineWeight-light nextui-table\\" aria-label=\\"Test sorting table\\" id=\\"react-aria9575136216-15\\" role=\\"grid\\" tabindex=\\"0\\" aria-describedby=\\"\\"><thead class=\\"nextui-c-PJLV nextui-table-row-group\\" role=\\"rowgroup\\"><tr class=\\"nextui-c-fRbcxI nextui-table-header-row\\" role=\\"row\\"><th class=\\"nextui-c-jhJWUH nextui-c-PJLV nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-column-header\\" tabindex=\\"-1\\" data-key=\\".0.0\\" role=\\"columnheader\\" id=\\"react-aria9575136216-15-.0.0\\">NAME</th><th class=\\"nextui-c-jhJWUH nextui-c-PJLV nextui-c-jhJWUH-hHqiAG-allowsSorting-true nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-column-header\\" tabindex=\\"-1\\" data-key=\\".0.1\\" role=\\"columnheader\\" id=\\"react-aria9575136216-15-.0.1\\" aria-sort=\\"none\\" aria-describedby=\\"react-aria-description-0\\">TYPE<svg aria-hidden=\\"true\\" class=\\"nextui-c-cqpNQI nextui-c-cqpNQI-ihOkgrg-css nextui-table-sort-icon\\" fill=\\"none\\" focusable=\\"false\\" height=\\"22\\" role=\\"presentation\\" viewBox=\\"0 0 24 24\\" width=\\"22\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M12 2C6.49 2 2 6.49 2 12C2 17.51 6.49 22 12 22C17.51 22 22 17.51 22 12C22 6.49 17.51 2 12 2ZM16.06 11.27L12.53 14.8C12.38 14.95 12.19 15.02 12 15.02C11.81 15.02 11.62 14.95 11.47 14.8L7.94 11.27C7.65 10.98 7.65 10.5 7.94 10.21C8.23 9.92 8.71 9.92 9 10.21L12 13.21L15 10.21C15.29 9.92 15.77 9.92 16.06 10.21C16.35 10.5 16.35 10.97 16.06 11.27Z\\"></path></svg></th><th class=\\"nextui-c-jhJWUH nextui-c-PJLV nextui-c-jhJWUH-hHqiAG-allowsSorting-true nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-column-header\\" tabindex=\\"-1\\" data-key=\\".0.2\\" role=\\"columnheader\\" id=\\"react-aria9575136216-15-.0.2\\" aria-sort=\\"none\\" aria-describedby=\\"react-aria-description-0\\">DATE MODIFIED<svg aria-hidden=\\"true\\" class=\\"nextui-c-cqpNQI nextui-c-cqpNQI-ihOkgrg-css nextui-table-sort-icon\\" fill=\\"none\\" focusable=\\"false\\" height=\\"22\\" role=\\"presentation\\" viewBox=\\"0 0 24 24\\" width=\\"22\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M12 2C6.49 2 2 6.49 2 12C2 17.51 6.49 22 12 22C17.51 22 22 17.51 22 12C22 6.49 17.51 2 12 2ZM16.06 11.27L12.53 14.8C12.38 14.95 12.19 15.02 12 15.02C11.81 15.02 11.62 14.95 11.47 14.8L7.94 11.27C7.65 10.98 7.65 10.5 7.94 10.21C8.23 9.92 8.71 9.92 9 10.21L12 13.21L15 10.21C15.29 9.92 15.77 9.92 16.06 10.21C16.35 10.5 16.35 10.97 16.06 11.27Z\\"></path></svg></th></tr><tr aria-hidden=\\"true\\" class=\\"nextui-c-gNVTSf nextui-c-gNVTSf-hakyQ-inline-false nextui-c-gNVTSf-icCJVMZ-css nextui-table-hidden-row\\"></tr></thead><tbody class=\\"nextui-c-PJLV nextui-c-PJLV-iccTTUn-css nextui-table-body\\" role=\\"rowgroup\\"><tr class=\\"nextui-c-bUbfzy nextui-c-PJLV nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-row\\" role=\\"row\\" tabindex=\\"-1\\" data-key=\\"1\\" aria-labelledby=\\"react-aria9575136216-15-1-.0.0\\" aria-hidden=\\"false\\"><td class=\\"nextui-c-PJLV nextui-c-czOCAM nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-cell\\" tabindex=\\"-1\\" data-key=\\"1.0\\" role=\\"rowheader\\" id=\\"react-aria9575136216-15-1-.0.0\\">Games</td><td class=\\"nextui-c-PJLV nextui-c-czOCAM nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-cell\\" tabindex=\\"-1\\" data-key=\\"1.1\\" role=\\"gridcell\\">File folder</td><td class=\\"nextui-c-PJLV nextui-c-czOCAM nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-cell\\" tabindex=\\"-1\\" data-key=\\"1.2\\" role=\\"gridcell\\">6/7/2020</td></tr></tbody></table></div>"`;

exports[`Table should render correctly 1`] = `"<div class=\\"nextui-c-hWZRae nextui-c-hWZRae-gikTHb-shadow-true nextui-c-hWZRae-bzEnZV-borderWeight-normal nextui-c-hWZRae-iPJLV-css nextui-table-container\\"><table class=\\"nextui-c-fcXuxO nextui-c-fcXuxO-fyxmJu-color-primary nextui-c-fcXuxO-fPgLZC-lineWeight-light nextui-table\\" aria-label=\\"Test example table\\" id=\\"react-aria9575136216-2\\" role=\\"grid\\" tabindex=\\"0\\" aria-describedby=\\"\\"><thead class=\\"nextui-c-PJLV nextui-table-row-group\\" role=\\"rowgroup\\"><tr class=\\"nextui-c-fRbcxI nextui-table-header-row\\" role=\\"row\\"><th class=\\"nextui-c-jhJWUH nextui-c-PJLV nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-column-header\\" tabindex=\\"-1\\" data-key=\\".0.0\\" role=\\"columnheader\\" id=\\"react-aria9575136216-2-.0.0\\">NAME</th><th class=\\"nextui-c-jhJWUH nextui-c-PJLV nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-column-header\\" tabindex=\\"-1\\" data-key=\\".0.1\\" role=\\"columnheader\\" id=\\"react-aria9575136216-2-.0.1\\">TYPE</th><th class=\\"nextui-c-jhJWUH nextui-c-PJLV nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-column-header\\" tabindex=\\"-1\\" data-key=\\".0.2\\" role=\\"columnheader\\" id=\\"react-aria9575136216-2-.0.2\\">DATE MODIFIED</th></tr><tr aria-hidden=\\"true\\" class=\\"nextui-c-gNVTSf nextui-c-gNVTSf-hakyQ-inline-false nextui-c-gNVTSf-icCJVMZ-css nextui-table-hidden-row\\"></tr></thead><tbody class=\\"nextui-c-PJLV nextui-c-PJLV-iccTTUn-css nextui-table-body\\" role=\\"rowgroup\\"><tr class=\\"nextui-c-bUbfzy nextui-c-PJLV nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-row\\" role=\\"row\\" tabindex=\\"-1\\" data-key=\\"1\\" aria-labelledby=\\"react-aria9575136216-2-1-.0.0\\" aria-hidden=\\"false\\"><td class=\\"nextui-c-PJLV nextui-c-czOCAM nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-cell\\" tabindex=\\"-1\\" data-key=\\"1.0\\" role=\\"rowheader\\" id=\\"react-aria9575136216-2-1-.0.0\\">Games</td><td class=\\"nextui-c-PJLV nextui-c-czOCAM nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-cell\\" tabindex=\\"-1\\" data-key=\\"1.1\\" role=\\"gridcell\\">File folder</td><td class=\\"nextui-c-PJLV nextui-c-czOCAM nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-table-cell\\" tabindex=\\"-1\\" data-key=\\"1.2\\" role=\\"gridcell\\">6/7/2020</td></tr></tbody></table></div>"`;
